File: extensions/vscode-editchain/src/multiplayer/progress.ts
/** Totals belong to a frozen inventory check, including already present records. */
export type CheckProgress = { pass: number; total_records: number | null; checked_records: number; complete: boolean; unavailable: number };
export type DownloadProgress = { content: boolean; received_bytes: number; total_bytes: number | null };
export type WorkProgress = { incoming?: CheckProgress; outgoing?: CheckProgress; pending_records?: number; pending_blobs?: number; download?: DownloadProgress | null };

const count = (value: unknown): value is number => Number.isSafeInteger(value) && (value as number) >= 0;
function validCheck(value: CheckProgress | undefined): boolean {
  return !!value && count(value.pass) && count(value.checked_records) && count(value.unavailable) && typeof value.complete === 'boolean'
    && (value.total_records === null ? !value.complete && value.checked_records === 0
      : count(value.total_records) && value.checked_records <= value.total_records && (!value.complete || value.checked_records === value.total_records));
}

export function validWorkProgress(value: WorkProgress): boolean {
  const download = value.download;
  return validCheck(value.incoming) && validCheck(value.outgoing) && count(value.pending_records) && count(value.pending_blobs)
    && (download === null || !!download && typeof download.content === 'boolean' && count(download.received_bytes)
      && (download.total_bytes === null ? download.received_bytes === 0 : count(download.total_bytes) && download.received_bytes <= download.total_bytes));
}

export function checking(value: WorkProgress & { synchronizing: boolean }): boolean {
  return value.synchronizing || !!value.outgoing?.pass && !value.outgoing.complete;
}

export function missingContent(value: WorkProgress & { unavailable: number }): boolean {
  return !!value.unavailable || !!value.outgoing?.unavailable;
}

/** Truncate so an unfinished check never rounds up to 100%. */
export function checkPercent(value?: CheckProgress): string | undefined {
  if (!value || value.total_records === null) return undefined;
  const percent = value.total_records === 0 ? (value.complete ? 100 : 0)
    : Math.floor(10000 * value.checked_records / value.total_records) / 100;
  return `${value.complete || value.total_records === 0 ? percent : Math.min(99.99, percent)}%`;
}

const number = (value: number) => value.toLocaleString('en-US');

export function describeCheck(label: string, value?: CheckProgress): string {
  if (!value || value.total_records === null) return `${label}: waiting for the shared-history total.`;
  const remaining = value.total_records - value.checked_records;
  const missing = value.unavailable ? ` ${number(value.unavailable)} content request(s) still unavailable.` : '';
  return `${label} check #${value.pass}: ${checkPercent(value)} — ${number(value.checked_records)}/${number(value.total_records)} records checked; ${number(remaining)} remaining to check.${missing}`;
}

export function describeDownload(value: WorkProgress): string {
  const queue = `Current receive batch: ${number(value.pending_records ?? 0)} records to save; ${number(value.pending_blobs ?? 0)} known content downloads left (includes the active one).`;
  const download = value.download;
  if (!download) return queue;
  const kind = download.content ? 'content' : 'record';
  if (download.total_bytes === null) return `${queue}\n  Waiting for ${kind} data; size not yet known.`;
  const percent = download.total_bytes === 0 ? 100 : Math.floor(1000 * download.received_bytes / download.total_bytes) / 10;
  return `${queue}\n  Downloading ${kind}: ${number(download.received_bytes)}/${number(download.total_bytes)} bytes (${percent}%); not yet saved.`;
}

/** Suppress routine completed pass numbers while still printing actual work. */
export function workSignature(value?: WorkProgress): unknown[] {
  const signature = (check?: CheckProgress) => check && [check.total_records, check.checked_records, check.complete, check.unavailable];
  return [signature(value?.incoming), signature(value?.outgoing), value?.pending_records, value?.pending_blobs, value?.download];
}

File: crates/editchain-sync/src/progress.rs
//! Credential-free synchronization work, separate from durable receipt counts.

use serde::Serialize;

/// One direction's stable inventory check. Checked records may have unavailable
/// content, so completing a check is not itself a claim of complete hydration.
#[derive(Debug, Default, Clone, Serialize)]
pub struct CheckProgress {
    /// Connection-local pass number; zero before the first request.
    pub pass: u64,
    /// Exact record count in the consent-filtered, frozen inventory.
    /// Absent until the first inventory page arrives.
    pub total_records: Option<u64>,
    /// Records whose page has finished checking records and referenced content.
    pub checked_records: u64,
    /// The last page has finished, including all requested content responses.
    pub complete: bool,
    /// Content requests that could not be satisfied in this pass.
    pub unavailable: u64,
}

/// One partially downloaded object. These bytes are not durable receipts.
#[derive(Debug, Clone, Serialize)]
pub struct DownloadProgress {
    /// Whether the object holds revision content rather than a record.
    pub content: bool,
    /// Bytes buffered locally, not yet validated and saved.
    pub received_bytes: u64,
    /// Length advertised by the first chunk; absent before that chunk.
    pub total_bytes: Option<u64>,
}

/// Current check work and durable, connection-local transfer counts.
#[derive(Debug, Default, Clone, Serialize)]
pub struct Progress {
    /// Space/version negotiation has succeeded over an authenticated channel.
    pub accepted: bool,
    /// An incoming reconciliation round is in progress.
    pub synchronizing: bool,
    /// Completed incoming inventory rounds, not a claim about future edits.
    pub rounds: u64,
    /// Exact records durably received on this connection.
    pub records: u64,
    /// Content blobs durably received on this connection.
    pub blobs: u64,
    /// Sent records acknowledged as durable by the authenticated remote peer.
    pub sent_records: u64,
    /// Sent blobs acknowledged as durable by the authenticated remote peer.
    pub sent_blobs: u64,
    /// Missing content responses in the current or last incoming round.
    pub unavailable: u64,
    /// This device's check of the peer's shared history.
    pub incoming: CheckProgress,
    /// The peer's confirmed check of this device's shared history.
    pub outgoing: CheckProgress,
    /// Known records in this page still to be saved, including staged bytes.
    pub pending_records: u64,
    /// Known content objects queued or being downloaded. More may be discovered.
    pub pending_blobs: u64,
    /// Buffered progress of the current object; never counted as saved data.
    pub download: Option<DownloadProgress>,
}
